consensus/consensus_tests/utils_test.go

```diff
@@ -304,17 +304,22 @@ func basePersistenceMock(t *testing.T, _ modules.EventsChannel) *modulesMock.Moc
 	ctrl := gomock.NewController(t)
 	persistenceMock := modulesMock.NewMockPersistenceModule(ctrl)
 	persistenceContextMock := modulesMock.NewMockPersistenceRWContext(ctrl)
+	persistenceReadContextMock := modulesMock.NewMockPersistenceReadContext(ctrl)
 
 	persistenceMock.EXPECT().Start().Return(nil).AnyTimes()
 	persistenceMock.EXPECT().SetBus(gomock.Any()).Return().AnyTimes()
-	persistenceMock.EXPECT().NewReadContext(int64(-1)).Return(persistenceContextMock, nil).AnyTimes()
+	persistenceMock.EXPECT().NewReadContext(gomock.Any()).Return(persistenceReadContextMock, nil).AnyTimes()
```

Comment:
https://github.com/pokt-network/pocket/pull/374#discussion_r1041605270

Reply:
This has been addressed. Now we are using the context from `Utility`

https://github.com/pokt-network/pocket/pull/378/commits/8f837201f2f2ec9132e72993bd98b1219cc2d88a

Reply:
I have a couple of observations I'd like to have your 2cents about: 
- is it by design that Utility always returns a RWContext and not a ReadOnly one?
- I find it weird that we instantiate the context passing in the height and then we have to specify the height again for all the calls like:

```go
	uCtx, err := m.GetBus().GetUtilityModule().NewContext(int64(message.Height))
	if err != nil {
		return typesCons.NodeId(0), err
	}
	vals, err := uCtx.GetPersistenceContext().GetAllValidators(int64(message.Height))
	if err != nil {
		return typesCons.NodeId(0), err
	}
```

Reply:
Great questions! tl;dr Legacy. Adding @gokutheengineer for context.

> is it by design that Utility always returns a RWContext and not a ReadOnly one?

No, I support adding the ability to instantiate a utility context with a read context.

> I find it weird that we instantiate the context passing in the height and then we have to specify the height again for all the calls like:

Agreed. I would say this:
1. A read context should be able to get instantiated without a height -> it only lets you read committed data on disk at any height.
2. A write context needs to be able to get instantiated with a height -> writes/update will happen, by default, at that height but not committed until the context is committed. Reads can happen for any height. Readds for he height provided for writing reads from the "context" (not committed.

@deblasis Can you create a ticket for this?

Reply:
> Agreed. I would say this:
> 
> 1. A read context should be able to get instantiated without a height -> it only lets you read committed data on disk at any height.
> 2. A write context needs to be able to get instantiated with a height -> writes/update will happen, by default, at that height but not committed until the context is committed. Reads can happen for any height. Readds for he height provided for writing reads from the "context" (not committed.

🤔 

## R/W

As far as I understand, a `RWContext` is supposed to be working in an append-only way. The state at a given `height` is immutable after it is committed / the block is finalized.
That said, it looks like the `RWContext` should not get the `height` argument in the constructor and also in the "setter" methods.
Why? 
Because we are writing a state where `height` = `currentheight`. 
Do we ever want to be able to overwrite a previous state? I don't think so but correct me if I am wrong.

## ReadOnly

For read only getters, it makes sense to have a height because technically you could be asking for a "view" of the state at any `height` since genesis.
The way I see the interface is that you instantiate providing a `height` and then all the getters from that context would be at that height. 

Let me know your thoughts on this.

Reply:
> Let me know your thoughts on this.

I think that you are completely, 💯 %, whole heartedly w̶r̶o̶n̶g̶ correct

tl;dr
- Write interface should accept a height
- Read interface should not accept a height
- Getters should accept a height
- Setters should not accept a height

🤝 ?

Reply:
😂 LOL

Ok, I just want to make sure I understand. I have two additional stupid questions:

>- Write interface should accept a height

1. If we can have only one RW context and we are working on the current block up until we commit it, why do we need a height? 

>- Read interface should not accept a height
>- Getters should accept a height

2. If the Read interface is meant to access the state at a certain height, why do we want to have to specify the height at every getter call and not once per-readcontext when we instantiate it?

>- Setters should not accept a height

👌

With these answers I guess I'll have enough context to write a mini-spec for the new ticket